src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -27,47 +27,92 @@ GcsActorScheduler::GcsActorScheduler(
     instrumented_io_context &io_context,
     GcsActorTable &gcs_actor_table,
     const GcsNodeManager &gcs_node_manager,
+    std::shared_ptr<ClusterTaskManager> cluster_task_manager,
     GcsActorSchedulerFailureCallback schedule_failure_handler,
     GcsActorSchedulerSuccessCallback schedule_success_handler,
     std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool,
-    rpc::ClientFactoryFn client_factory)
+    rpc::ClientFactoryFn client_factory,
+    std::function<void(const NodeID &, const rpc::ResourcesData &)>
+        normal_task_resources_changed_callback)
     : io_context_(io_context),
       gcs_actor_table_(gcs_actor_table),
       gcs_node_manager_(gcs_node_manager),
+      cluster_task_manager_(std::move(cluster_task_manager)),
       schedule_failure_handler_(std::move(schedule_failure_handler)),
       schedule_success_handler_(std::move(schedule_success_handler)),
       raylet_client_pool_(raylet_client_pool),
-      core_worker_clients_(client_factory) {
+      core_worker_clients_(client_factory),
+      normal_task_resources_changed_callback_(normal_task_resources_changed_callback) {
   RAY_CHECK(schedule_failure_handler_ != nullptr && schedule_success_handler_ != nullptr);
 }
 
 void GcsActorScheduler::Schedule(std::shared_ptr<GcsActor> actor) {
   RAY_CHECK(actor->GetNodeID().IsNil() && actor->GetWorkerID().IsNil());
 
-  // Select a node to lease worker for the actor.
-  auto node_id = SelectNode(actor);
-
-  auto node = gcs_node_manager_.GetAliveNode(node_id);
-  if (!node.has_value()) {
-    // There are no available nodes to schedule the actor, so just trigger the failed
-    // handler.
-    schedule_failure_handler_(std::move(actor),
-                              rpc::RequestWorkerLeaseReply::SCHEDULING_FAILED,
-                              "No available nodes to schedule the actor");
-    return;
+  if (RayConfig::instance().gcs_actor_scheduling_enabled()) {
+    ScheduleByGcs(actor);
+  } else {
+    ScheduleByRaylet(actor);
   }
+}
+
+void GcsActorScheduler::ScheduleByGcs(std::shared_ptr<GcsActor> actor) {
+  auto callback = [this, actor](const NodeID &node_id) {
+    auto node = gcs_node_manager_.GetAliveNode(node_id);
+    RAY_CHECK(node.has_value());
+
+    // Update the address of the actor as it is tied to a node.
+    rpc::Address address;
+    address.set_raylet_id(node.value()->node_id());
+    actor->UpdateAddress(address);
+
+    RAY_CHECK(node_to_actors_when_leasing_[actor->GetNodeID()]
+                  .emplace(actor->GetActorID())
+                  .second);
+
+    auto &acquired_resources = actor->GetMutableAcquiredResources();
+    acquired_resources = ResourceMapToResourceRequest(
+        actor->GetCreationTaskSpecification().GetRequiredResources().GetResourceMap(),
+        false);
```

Comment:
Why do we need this code? It doesn't exist for `ScheduleByRaylet`

Reply:
We use this to record how much resources we acquired from gcs's cluster_resource_manager. For ScheduleByRaylet, we acquire nothing because actor is directly forwarded to a remote node. 

When destroying or rescheduling an actor, we then use this field to return acquired resources (for Raylet scheduler, this field is empty, so we return nothing).